#!/bin/bash

### Script to run htseq_count after running RNA-Seq alignments using Bowtie2

### Job name
#PBS -N htseq_count

### Specify email address to use for notification.
#PBS -M bmlau@email.arizona.edu

### Request email when job ends
#PBS -m ea

### Specify the PI group found with va command
#PBS -W group_list=mfh

### Set the queue to submit this job.
#PBS -q standard

### Set the number of nodes and cpus that will be used.
#PBS -l select=1:ncpus=5:mem=30gb:pcmem=6gb
#PBS -l place=pack:shared

### Specify up to a maximum of 1600 hours total cpu time
#PBS -l cput=020:00:0

### Specify up to a maximum of 240 hours walltime for the job
#PBS -l walltime=10:00:00

### Define path for stdout and stderr
#PBS -o ../../../Processing/Logs
#PBS -j oe

### Move to *current* directory
cd $PBS_O_WORKDIR

module load samtools
module load python/2

### Set log file name
LOGFILE=$PBS_O_WORKDIR/../../../Processing/Logs/$PBS_JOBID.htseq_count.OU

### Profile job
echo Running on host `hostname` > $LOGFILE
echo Time is `date` >> $LOGFILE
echo Directory is `pwd` >> $LOGFILE
echo >> $LOGFILE

TODAY=`date +%d%b%Y`
OUTPUT_DIR=${PBS_O_WORKDIR}/htseq_count.${TODAY}
mkdir -p ${OUTPUT_DIR}

### Run job
cd ${PBS_O_WORKDIR}

mkdir -p ${OUTPUT_DIR}
htseq-count \
-f bam \
-m union \
-s no \
-i gene_id \
Aligned.out.bam \
/rsgrps/exome/bmlau/RNA-Seq/STAR/Indices/Human/Gencode_v19/gencode.v19.annotation.collapsed.gtf \
> ${OUTPUT_DIR}/counts.txt 2>> $LOGFILE

### Get time stamp again
date >> $LOGFILE


